Introduced per connection gate to control connection life cycle

mmaslankaprv opened the pull request:
## Cover letter

Previously in Redpanda `net::server` we used a single gate that controlled life cycle of  all the connections. However when one of the connections input stream is broken or redpanda is shutting down (in other words `protocol::apply` method returns) connection must be closed. In current implementation `connection::shutdown()` member function doesn't wait for all the connection related processing to finish. This may lead to situation in which connection output stream is closed before handler finished processing request and sending a response. This may lead to the situation in which operation was successful but redpanda returned an error or failed to return any answer.

Fixed this problem by introducing a per connection gate that keeps the connection alive until all the pending requests are finished.   

Related: #3277 

<!-- Use the GitHub keyword `Fixes` to link to bug(s) this PR will fix. -->
Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

## Release notes
<!--

If this PR does not need to be included in the release notes, then
simply have a bullet point for `none` directly under the `Release notes`
section, e.g.

* none

Otherwise, add one or more of the following sections. A section must have
at least 1 bullet point. You can add multiple sections with multiple
bullet points if this PR represents multiple release note items. See
the CONTRIBUTING.md guidelines for more details.

### Features

* Short description of the feature. Explain how to configure the new feature if applicable.

### Improvements

* Short description of how this PR improves redpanda.

-->

dotnwat reviewed (commented):
I have a few questions:

Does this solve a general problem? For example, in the kafka layer we launch the background handling of a request under the connection gate, but we still interact with the connection before that. So I don't think this would solve the general problem of interacting with a shutdown connection, unless I'm missing something.

Can this result in a deadlock? For example, one thing I assumed that `_out.close` would do is force any operations on the socket that are blocked to be forcefully unblocked at the OS level. If an operation is stuck, and we want on the gate before closing, then would that result in a deadlock?

Is the issue we are seeing an assertion failure or an exception being thrown (or something else)? If it's an exception being thrown, then it seems like the general solution is to evaluate some of our paths to handle exceptions more intelligently?

How should I now think about net::server::_conn_gate? Should it still be involved in determinations about if a reply can be sent?

Should we drop the net::logger patch and let the server specializations (e.g. rpc or kafka) provide a logger object? That was the original intent, but we can also at net:logger for now and adjust this later.

mmaslankaprv commented:
> I have a few questions:
> 
> Does this solve a general problem? For example, in the kafka layer we launch the background handling of a request under the connection gate, but we still interact with the connection before launching that background fiber. So I don't think this would solve the general problem of interacting with a shutdown connection because the shutdown could still be interleaved with interacting with the connection where those interactions aren't under the gate. Should they be under the gate even though they aren't necessarily in the background?

i think you are right, in the first version of this i've added the gate_holder to `kafka::connection_context` but then i've removed it to minimize code churn. 

> 
> Can this result in a deadlock? For example, one thing I assumed that `_out.close` would do is force any operations on the socket that are blocked to be forcefully unblocked at the OS level. If an operation is stuck, and we want on the gate before closing, then would that result in a deadlock?

I wouldn't necessary called this a deadlock, indeed even before this PR, when `connection::shutdown` is called we just close the `_out` stream, this does not influence any of connection related pending requests, moreover even before this PR the connection object may be kept alive indefinitely after connection wash shut down. The change that this PR introduces would not allow connection shutdown when there are pending requests explicitly not like it is now. A

> 
> Is the issue we are seeing an assertion failure or an exception being thrown (or something else)? If it's an exception being thrown, then it seems like the general solution is to evaluate some of our paths to handle exceptions more intelligently?

The problem is that when redpanda is requested to stop gracefully it will not be able to finish its work since connections are shutdown before background operation finished. There is no additional cost of waiting for those to finish as they must finish since we are eventually waiting for all the gates. The difference is that with this PR we give handler a chance to send generated reply back to requester before `_out` stream is closed by the same time guaranteeing that no new data can be received on this connection.  

> 
> How should I now think about net::server::_conn_gate? Should it still be involved in determinations about if a reply can be sent?
> 
> Should we drop the net::logger patch and let the server specializations (e.g. rpc or kafka) provide a logger object? That was the original intent, but we can also at net:logger for now and adjust this later.

Good idea, i will change that after we settle on the other things.

mmaslankaprv commented:
@dotnwat we were chatting about this PR some time last week, do you have any other comments ?

jcsp commented:
This could be problematic if the remote end of a connection is not accepting data -- any writes into the output stream can be blocked indefinitely, and with this change we are waiting for the request processing (including those blocking writes) before we shut down.

To accomplish the goal of a clean shutdown completing in-flight requests under happy conditions, while avoiding hangs, I think we'd need a timeout.  Initially close the gate, then if all gate holders haven't released after T, then call shutdown_input to prevent a hang.

dotnwat reviewed (commented):
@mmaslankaprv I don't want to let this PR drop--I think it does address something important--but at the same time the PR isn't merged and we don't seem to be having problems.

Is there an urgency is merging this, and if not, is there a solution you would prefer that may take more effort that wasn't originally considered? Also I'm happy to review this again with fresh eyes.
